earth_enterprise/src/common/khMTTypes.h

```diff
@@ -148,9 +151,166 @@ class khMTSet
     khLockGuard guard(mutex);
     return std::vector<T>(set.begin(), set.end());
   }
+  void clear(void) {
+    khLockGuard guard(mutex);
+    set.clear();
+  }
+  size_t size(void) {
+    return set.size();
+  }
+};
+
+
+template <class K, class V>
+class khMTMap
+{
+ protected:
+  typedef std::unordered_map<K,V> Map;
+  Map     map;
+  mutable std::mutex mutex;
+
+ private:
+  // private and unimplimented
+  khMTMap& operator=(const khMTMap&) = delete;
+
+  khMTMap(const khMTMap& o, const std::lock_guard<std::mutex>&) : map(o.map) { }
+  khMTMap(khMTMap&& o, const std::lock_guard<std::mutex>&) : map(std::move(o.map)) { }
+ public:
+  khMTMap(void) { }
+  khMTMap(const khMTMap& o) : khMTMap(o, std::lock_guard<std::mutex>(o.mutex)) { }
+  khMTMap(khMTMap&& o) : khMTMap(o, std::lock_guard<std::mutex>(o.mutex)){ }
+
+  void insert(const std::pair<K,V> &val) {
+    std::lock_guard<std::mutex> guard(mutex);
+    (void)map.insert(val);
+  }
+  void erase(const K &val) {
+    std::lock_guard<std::mutex> guard(mutex);
+    (void)map.erase(val);
+  }
+  size_t size(void) {
+    return map.size();
+  }
+  V& operator[](const K& key) {
+    std::lock_guard<std::mutex> guard(mutex);
+    return map[key];
+  }
+  void clear(void) {
+    std::lock_guard<std::mutex> guard(mutex);
+    map.clear();
+  }
+  std::vector<K> keys(void) const {
+    std::lock_guard<std::mutex> guard(mutex);
+    std::vector<K> thekeys;
+    thekeys.reserve(map.size());
+    for (const auto& p : map) {
+      thekeys.push_back(p.first);
+    }
+    return thekeys;
+  }
+  bool contains(const K& key) {
+    std::lock_guard<std::mutex> guard(mutex);
+    return map.find(key) != map.end();
+  }
+  bool empty(void) const {
+    std::lock_guard<std::mutex> guard(mutex);
+    return map.empty();
+  }
 };
 
+// WARNING: Use this with extreme caution
+// It protects some obvious problems but creates some less obvious ones
+template <typename T>
+class MTVector {
+  public:
+    typedef std::vector<T> Base;
+    typedef typename Base::const_iterator const_iterator;
 
+  private:
+    mutable std::mutex mtx;
```

Comment:
Note for reviewers: we could change this to be a pthread read/write lock.  From my testing on an 8 core machine there was minimal benefit to doing this, so I left it out.

Reply:
Is there any significant downside? It might be nice to have the code just in case.